paddle/framework/net_design.md

```diff
@@ -0,0 +1,249 @@
+# Network Design
+
+`Network` is the container and controller of a set of operators,
+users can build a real network from a `NetDef` which is a protobuf message 
+and use `Network.Run()` to run all the operators in the network.
+
+The `Network` will
+
+- manage all the operators contained in the network.
+- not own any `Variable`.
+
+# API
+
+## NetBase
+To make the `Network` extendable, a base class is defined like this
+
+```c++
+// operator's index stored in a network.
+typedef int OpIndex;
+
+// The minimum a network should be implemented.
+class NetBase {
+ public:
+  // run all the operators and return success(true) or not, all the
+  // variables are located in `scope`. `begin` and `end` specify the scope of
+  // `ops_` to run, If no positive indexes are provided, all operators in `ops_`
+  // will run.
+  virtual bool Run(Scope *scope, OpIndex begin = -1,
+                   OpIndex end = -1) const = 0;
+
```

Comment:
Actually, the NetBase Run method is what an engine should do. The engine will get variables from scope, and get runtime resources from context, and then execute.
The runtime resources may include cuda stream, cublasHandle, cudnnHandle.